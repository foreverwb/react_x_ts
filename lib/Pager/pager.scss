@import "../_main";
%item {
  height: $input-height;
  width: $input-height;
  border: 1px solid $input-border-color;
  background: none;
  border-radius: $input-border-radius;
  margin: 0 4px;
  vertical-align: center;
}
%center {
  display: flex;
  align-content: center;
  justify-content: center;
}
.ui-pager {
  display: flex;
  margin: 8px 0;
  &-item {
    @extend %item;
  }
  &-prev {
    @extend %item; @extend %center
  }
  &-next {
    @extend %item; @extend %center
  }
}