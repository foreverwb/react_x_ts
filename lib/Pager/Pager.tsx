import React from 'react';
import {scopedClassMaker} from '../untils/classes';
import Icon from '../Icon/Icon';
import './pager.scss';

const sc = scopedClassMaker('ui-pager');

function range(start: number, end: number): number[] {
  const result: number[] = [];
  for (let i = start; i <= end; i++) {
    result.push(i);
  }
  return result;
}

interface IProps {
  current?: number;
  total: number;
  defaultCurrent?: number;
  onChange: (value: number) => void
}

interface IState {
  current?: number
}

class Pager extends React.Component<IProps, IState> {
  static defaultProps = {};

  get current() {
    if ('defaultCurrent' in this.props) {
      return this.state.current || 1;
    } else {
      return this.props.current || 1;
    }
  }

  set current(value) {
    if ('defaultCurrent' in this.props) {
      this.setState({current: value});
    } else {
      this.props.onChange && this.props.onChange(value);
    }
  }

  onClickItem = (value: number, e: React.MouseEvent<HTMLSpanElement>) => {
    if (value <= this.props.total && value >= 1) {
      this.current = value;
    }
  };

  jumpPage = (index: number) => {
    const prev = this.current - 5 <= 0 ? 1 : this.current - 5;
    const next = this.current + 5 >= this.props.total ? this.props.total : this.current + 5;
    return index === 1 ? prev : next;
  };

  get items() {
    return range(1, this.props.total)
      .filter(item => {
        if (item === 1) return true;
        if (item === 2) return true;
        else if (item === this.props.total) return true;
        else return Math.abs(item - this.current) <= 2;
      })
      .reduce((prev, next) => {
        const last = prev[prev.length - 1];
        const temp = last !== -1 && last - next < -1;
        return prev.concat(temp ? [-1, next] : [next]);
      }, [] as number[])
      .map((entry, idx) => entry === -1 ?
        <span
          key={idx}
          onClick={e => this.onClickItem(this.jumpPage(idx), e)}
        >
          ...
        </span> :
        <button
          key={idx}
          className={sc('item')}
          onClick={e => this.onClickItem(entry, e)}
        >
          {entry}
        </button>
      );

  }

  render() {
    return (
      <div className={sc('')}>
        <button
          className={sc('prev')}
          onClick={this.onClickItem.bind(null, this.current - 1)}
        >
          <Icon name="left"/>
        </button>
        {this.items}
        <button
          className={sc('next')}
          onClick={this.onClickItem.bind(null, this.current + 1)}
        >
          <Icon name="right"/>
        </button>
      </div>
    );
  }
}

export default Pager;