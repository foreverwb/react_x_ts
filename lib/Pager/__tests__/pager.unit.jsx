import React from 'react';
import renderer from 'react-test-renderer';
import Pager from '../Pager';
import {mount} from 'enzyme';

describe('Pager', () => {
  it('prop current = 1', () => {
    const temp = renderer
      .create(<Pager current={1} total={3}/>)
      .toJSON();
    expect(temp).toMatchSnapshot();
  });
  it('prop total = 5', () => {
    const temp = renderer
      .create(<Pager total={5}/>)
      .toJSON();
    expect(temp).toMatchSnapshot();
  });
  it('应该有两个省略号', () => {
    const temp = renderer
      .create(<Pager total={20} current={6}/>)
      .toJSON();
    expect(temp).toMatchSnapshot();
  });
  it('点击对应按钮current发生对应的变化', () => {
    const c = mount(
      <Pager
        current={1}
        total={5}
        onChange={(value) => c.setProps({current: value})}
      />
    );
    expect(c.find('.ui-pager-active').text()).toEqual('1');
    c.find('.ui-pager-item').last().simulate('click');
    expect(c.find('.ui-pager-active').text()).toEqual('5');
  });
});