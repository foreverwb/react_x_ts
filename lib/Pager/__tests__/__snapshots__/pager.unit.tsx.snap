// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pager prop current = 1 1`] = `
<div
  className="ui-pager"
>
  <button
    className="ui-pager-prev"
    onClick={[Function]}
  >
    <svg
      className="ui-icon"
    >
      <use
        xlinkHref="#left"
      />
    </svg>
  </button>
  <button
    className="ui-pager-item ui-pager-active"
    onClick={[Function]}
  >
    1
  </button>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    2
  </button>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    3
  </button>
  <button
    className="ui-pager-next"
    onClick={[Function]}
  >
    <svg
      className="ui-icon"
    >
      <use
        xlinkHref="#right"
      />
    </svg>
  </button>
</div>
`;

exports[`Pager prop total = 5 1`] = `
<div
  className="ui-pager"
>
  <button
    className="ui-pager-prev"
    onClick={[Function]}
  >
    <svg
      className="ui-icon"
    >
      <use
        xlinkHref="#left"
      />
    </svg>
  </button>
  <button
    className="ui-pager-item ui-pager-active"
    onClick={[Function]}
  >
    1
  </button>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    2
  </button>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    3
  </button>
  <span
    onClick={[Function]}
  >
    ...
  </span>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    5
  </button>
  <button
    className="ui-pager-next"
    onClick={[Function]}
  >
    <svg
      className="ui-icon"
    >
      <use
        xlinkHref="#right"
      />
    </svg>
  </button>
</div>
`;

exports[`Pager 应该有两个省略号 1`] = `
<div
  className="ui-pager"
>
  <button
    className="ui-pager-prev"
    onClick={[Function]}
  >
    <svg
      className="ui-icon"
    >
      <use
        xlinkHref="#left"
      />
    </svg>
  </button>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    1
  </button>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    2
  </button>
  <span
    onClick={[Function]}
  >
    ...
  </span>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    4
  </button>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    5
  </button>
  <button
    className="ui-pager-item ui-pager-active"
    onClick={[Function]}
  >
    6
  </button>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    7
  </button>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    8
  </button>
  <span
    onClick={[Function]}
  >
    ...
  </span>
  <button
    className="ui-pager-item"
    onClick={[Function]}
  >
    20
  </button>
  <button
    className="ui-pager-next"
    onClick={[Function]}
  >
    <svg
      className="ui-icon"
    >
      <use
        xlinkHref="#right"
      />
    </svg>
  </button>
</div>
`;
